### ISSUE TITLE
Does whitebox XFR methods support PReLU layers and InsightFace IResNet models?

### ISSUE THREAD

**Issue by manideep2510** [stresearch/xfr#9]
Hi @jebyrne and @williford 

I am trying to use EBP visualization on [InsightFace's IResNet models](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_torch/backbones/iresnet.py) which contain one non-trivial layer, i.e, PReLU layers instead of plain ReLU layers. I could not get visualizations using these trained IResNet models (giving 512d embeddings for input faces). The saliency map is coming out to be NaNs. Initial 6-7 gradients (of the last 6-7 layers) are okay, but after that, they are becoming NaNs. 

I have also tried replacing PReLUs with normal ReLU layers and used random initialization of the model, replaced operations like `+` and `*` with `Add()` and `Multiply()` PyTorch nn.Module classes as recommended in the readme, but I still get NaNs as output.

Can you please help me debug this? Thanks in advance.

**Comment by williford** [comment]
Hi Manideep, have you made any progress in solving this?

The negative outputs from PReLU could cause a problem, although I would have thought swapping to ReLU would have fixed it. Random initialization might be a problem, because the activation values would become so small? I'm not familiar with the current Pytorch implementation of EBP, but I would expect that you could do a forward pass as is, then when you are doing the backward pass you would need to replace the negative activations with zero.

**Comment by manideep2510** [comment]
Hi @williford. Thank you for your response

I couldn't make this work yet. I have re-trained the IResNet50 model from InsightFace by replacing PReLU layers with ReLU layers. And the loaded this model for EBP saliency map visualization, but this also does not work. 

Interestingly using the [ResNet model from your repository](https://github.com/stresearch/xfr/blob/master/python/xfr/models/resnet.py) (with random initialization) in Whitebox visualization code gives good saliency map outputs. You have used this network originally for the BlackBox visualizations. After replacing PReLU with ReLU, when I compare the [ResNet code you have provided](https://github.com/stresearch/xfr/blob/master/python/xfr/models/resnet.py) and the [code in the InsightFace repository](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_torch/backbones/iresnet.py), I couldn't find any differences or deviations between them except the order of convolutions in the ResNet blocks, but strangely the Insightface version on ResNet still give NaN outputs for saliency maps.

Please let me know if you can find any differences between these codebases apart from the activations, which might be causing the issue. I will try to train your ResNet implementation on our data and will verify if it works.

**Comment by manideep2510** [comment]
One more difference between these two implementations is that the InsightFace implementation uses `torch.flatten()` to flatten out the feature maps from the last Conv layer instead of the AveragePooling layer used in the ResNet implementation you have provided. Could this be a problem, in your opinion?

**Comment by manideep2510** [comment]
Hi

I wanted to let you know that I was able to solve this issue. The problem had nothing to do with the type of network in question. I was making a small mistake (which took me a long time to detect 😅 ) because of which the gradients are not being calculated correctly. 

Here's the mistake I was making. In [whitebox.py](https://github.com/stresearch/xfr/blob/master/python/xfr/models/whitebox.py), in [line 18](https://github.com/stresearch/xfr/blob/906c393b8a95098a498fd24aea149186ebb7f207/python/xfr/models/whitebox.py#L18), `Bottleneck` is being imported from the same file as that of the ResNet model. This imported `Bottleneck` module is required to transverse the layers properly by the [`_layer_visitor`](https://github.com/stresearch/xfr/blob/906c393b8a95098a498fd24aea149186ebb7f207/python/xfr/models/whitebox.py#L34) function, and importing this module from the correct file is needed, otherwise `_layer_visitor` will not be able to access the layers inside the `Bottleneck` module properly. But, I did not know about this, and I was importing the ResNet model from one script and the `Bottleneck` from a different script. The visualizations worked as expected upon correcting this, i.e., importing the `Bottleneck` from the same script as that of the ResNet (i.e., the InsightFace code).

Thanks a lot for the help.
